src/JustEat.StatsD/SimpleObjectPool.cs

```diff
@@ -1,56 +1,58 @@
-    public sealed class SimpleObjectPool<T>
-        where T : class
+    /// <summary>
+    /// A class that provides simple thread-safe object pooling semantics.
+    /// </summary>
+    public sealed class SimpleObjectPool<T> where T : class
     {
+        private readonly Func<SimpleObjectPool<T>, T> _constructor;
         private readonly ConcurrentBag<T> _pool;
 
         /// <summary>	Constructor that populates a pool with the given number of items. </summary>
-        /// <exception cref="ArgumentNullException">	Thrown when the constructor is null. </exception>
-        /// <exception cref="ArgumentException">		Thrown when the constructor produces null objects. </exception>
-        /// <param name="capacity">   	The capacity. </param>
-        /// <param name="constructor">	The factory method used to create new instances of the object to populate the pool. </param>
-        [SuppressMessage("Microsoft.Design", "CA1006:DoNotNestGenericTypesInMemberSignatures", Justification = "Necessary to nest generics to support passing a factory method")]
-        public SimpleObjectPool(int capacity, Func<SimpleObjectPool<T>, T> constructor)
+        /// <exception cref="ArgumentNullException"> Thrown when the constructor is null. </exception>
+        /// <param name="constructor"> The factory method used to create new instances of the object to populate the pool. </param>
+        public SimpleObjectPool(int initialSize, Func<SimpleObjectPool<T>, T> constructor)
         {
-            if (constructor == null)
-            {
-                throw new ArgumentNullException(nameof(constructor));
-            }
-
+            _constructor = constructor ?? throw new ArgumentNullException(nameof(constructor));
             _pool = new ConcurrentBag<T>();
-            for (var i = 0; i < capacity; ++i)
+            PrePopulate(initialSize);
+        }
+
+        private void PrePopulate(int initialSize)
+        {
+            for (var i = 0; i < initialSize; ++i)
             {
-                var instance = constructor(this);
+                var instance = _constructor(this);
                 if (instance == null)
                 {
-                    throw new ArgumentException("constructor produced null object", "constructor");
+                    throw new ArgumentException("constructor produced null object");
                 }
 
                 _pool.Add(instance);
             }
         }
 
-        /// <summary>	Retrieves an object from the pool if one is available. </summary>
-        /// <returns>	An object or null if the pool has been exhausted. </returns>
+        public int Count => _pool.Count;
+
+        /// <summary>Retrieves an object from the pool if one is available.
+        /// Expand the pool if need be </summary>
+        /// <returns>An object from the pool. </returns>
         public T Pop()
         {
-            if (!_pool.TryTake(out T result))
+            if (_pool.TryTake(out T result))
```

Comment:
I think _technically_ this is a breaking change if anyone is using this class, as before you could use `null` to mean _is empty_ like when a `StreamReader` returns `null`. Now you can't, and you'd just run in a loop forever creating objects.

Reply:
Yep, that exactly what the test did :(

Reply:
There is now `Pop` and `PopOrCreate`